exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/scan/v3/schema/ScanSchemaResolver.java

```diff
@@ -189,7 +189,7 @@ private void insertColumn(ColumnMetadata col) {
     switch (mode) {
       case FIRST_READER_SCHEMA:
       case READER_SCHEMA:
-        if (schema.projectionType() != ProjectionType.ALL) {
+        if (schema.projectionType() != ProjectionType.ALL && !col.isArray()) {
```

Comment:
Why is an array column special here? I think this is trying to say that, if the array is not projected, it should not have been created. There are dummy structures used instead. This fix suggests that there is a bug somewhere other than here.

Reply:
Let's look at this case :
1. batch reader received the project column, from ` SELECT path, data_type, file_name FROM ` in the constructor().
2. schema is [TupleSchema [ProjectedColumn [`path` (LATE:REQUIRED)]], [ProjectedColumn [`data_type` (LATE:REQUIRED)]], [ProjectedColumn [`file_name` (LATE:REQUIRED)]]].
3. ProjectionType = 'SOME'.
4. batch reader create new array column in next().
5. do the project for the schema.
```java
ColumnHandle existing = schema.find(colSchema.name());
 if (existing == null) {
  insertColumn(colSchema);
}
```
6. catch and throw the `IllegalStateException`.
7. failed to reader the format data.

In EVF1, creating array fields dynamically is allowed and does not throw such exceptions.

Reply:
This error check is correct: the problem must be in the `ResultSetLoader`. This error says that the `ResultSetLoader` is trying to add a materialized column, but that column is not projected. The result set loader should have created a dummy column and not passed the column along to this mechanism.